components/places/src/storage/history.rs

```diff
@@ -188,6 +188,144 @@ pub fn delete_place_by_guid(db: &impl ConnExt, guid: &SyncGuid) -> Result<()> {
 }
 
+/// Delete all visits in a date range.
+pub fn delete_visits_since(db: &impl ConnExt, since: Timestamp) -> Result<()> {
+    let tx = db.unchecked_transaction()?;
+    delete_visits_since_in_tx(db, since)?;
+    tx.commit()?;
+    Ok(())
+}
+
+pub fn delete_visits_since_in_tx(db: &impl ConnExt, since: Timestamp) -> Result<()> {
+    // Like desktop's removeVisitsByFilter, we query the visit and place ids
+    // affected, then delete all visits, then delete all place ids in the set
+    // which are orphans after the delete.
+    let sql = "SELECT id, place_id FROM moz_historyvisits WHERE visit_date >= :since";
+    let visits =
+        db.query_rows_and_then_named(sql, &[(":since", &since)], |row| -> rusqlite::Result<_> {
+            Ok((
+                row.get_checked::<_, RowId>(0)?,
+                row.get_checked::<_, RowId>(1)?,
+            ))
+        })?;
+
+    sql_support::each_chunk_mapped(
+        &visits,
+        |(visit_id, _)| visit_id,
+        |chunk, _| -> Result<()> {
+            db.conn().execute(
+                &format!(
+                    "DELETE from moz_historyvisits WHERE id IN ({})",
+                    sql_support::repeat_sql_vars(chunk.len()),
+                ),
+                chunk,
+            )?;
+            Ok(())
+        },
+    )?;
+
+    // Find out which pages have been possibly orphaned and clean them up.
+    sql_support::each_chunk_mapped(
+        &visits,
+        |(_, place_id)| place_id.0,
+        |chunk, _| -> Result<()> {
+            let query = format!(
+                "SELECT id, -- url, url_hash, guid
+                    (foreign_count != 0) AS has_foreign,
+                    ((last_visit_date_local + last_visit_date_remote) != 0) as has_visits
+                FROM moz_places
+                WHERE id IN ({})",
+                sql_support::repeat_sql_vars(chunk.len()),
+            );
+
+            let mut stmt = db.conn().prepare(&query)?;
+            let page_results = stmt.query_and_then(chunk, PageToClean::from_row)?;
+            let pages: Vec<PageToClean> = page_results.collect::<Result<_>>()?;
+            cleanup_pages(db, &pages)
+        },
+    )?;
+
+    Ok(())
+}
+
+#[derive(Debug)]
+struct PageToClean {
+    id: RowId,
+    has_foreign: bool,
+    has_visits: bool,
+}
+
+impl PageToClean {
+    pub fn from_row(row: &Row) -> Result<Self> {
+        Ok(Self {
+            id: row.get_checked("id")?,
+            has_foreign: row.get_checked("has_foreign")?,
+            has_visits: row.get_checked("has_visits")?,
+        })
+    }
+}
+
+/// Clean up pages whose history has been modified, by either
+/// removing them entirely (if they are marked for removal,
+/// typically because all visits have been removed and there
+/// are no more foreign keys such as bookmarks) or updating
+/// their frecency.
+fn cleanup_pages(db: &impl ConnExt, pages: &[PageToClean]) -> Result<()> {
+    // desktop does this frecency work using a function in a single sql
```

Comment:
Eh, it's worth noting that desktop's implementation (like ours) still needs to run rust code which makes a number of internal SQL queries, so the only real difference is that SQLite is the one who makes the call and does the loop. We shouldn't expect SQLite to do either of those better than us.

In fact, a direct call like we make here is going to run (slightly) faster than an indirect call that SQLite makes (let alone the fact that rusqlite has to wrap it in both `Box<FnMut(...)>` and `catch_unwind` to avoid unwinding across the FFI boundary...). So I'd actually expect doing it our way to be faster than how desktop does it. (Not to mention, We actually will propagate errors correctly across this boundary, whereas rusqlite will type erase them).